src/Mod/Draft/draftobjects/patharray.py

```diff
@@ -0,0 +1,272 @@
+
+        obj.Count = 2
+        obj.PathSubs = []
+        obj.Xlate = App.Vector(0,0,0)
+        obj.Align = False
+        obj.TangentVector = App.Vector(1.0, 0.0, 0.0)
+
+        if self.use_link:
+            _tip = "Show array element as children object"
+            obj.addProperty("App::PropertyBool","ExpandArray",
+                            "Parameters", QT_TRANSLATE_NOOP("App::Property", _tip))
+
+            obj.ExpandArray = False
+            obj.setPropertyStatus('Shape','Transient')
+
+        super(PathArray, self).attach(obj)
+
+    def linkSetup(self,obj):
+        super(PathArray, self).linkSetup(obj)
+        obj.configLinkProperty(ElementCount='Count')
+
+    def execute(self,obj):
+        import Part
+        import DraftGeomUtils
+        if obj.Base and obj.PathObj:
+            pl = obj.Placement
+            if obj.PathSubs:
+                w = self.getWireFromSubs(obj)
+            elif (hasattr(obj.PathObj.Shape,'Wires') and obj.PathObj.Shape.Wires):
+                w = obj.PathObj.Shape.Wires[0]
+            elif obj.PathObj.Shape.Edges:
+                w = Part.Wire(obj.PathObj.Shape.Edges)
+            else:
+                App.Console.PrintLog ("_PathArray.createGeometry: path " + obj.PathObj.Name + " has no edges\n")
+                return
+            if (hasattr(obj, "TangentVector")) and (obj.Align):
+                basePlacement = obj.Base.Shape.Placement
+                baseRotation = basePlacement.Rotation
+                stdX = App.Vector(1.0, 0.0, 0.0)
+                preRotation = App.Rotation(stdX, obj.TangentVector)   #make rotation from X to TangentVector
+                netRotation = baseRotation.multiply(preRotation)
+                base = calculatePlacementsOnPath(
+                        netRotation,w,obj.Count,obj.Xlate,obj.Align)
+            else:
+                base = calculatePlacementsOnPath(
+                        obj.Base.Shape.Placement.Rotation,w,obj.Count,obj.Xlate,obj.Align)
+            return super(PathArray, self).buildShape(obj, pl, base)
+
+    def getWireFromSubs(self,obj):
+        '''Make a wire from PathObj subelements'''
+        import Part
+        sl = []
+        for sub in obj.PathSubs:
+            edgeNames = sub[1]
+            for n in edgeNames:
+                e = sub[0].Shape.getElement(n)
+                sl.append(e)
+        return Part.Wire(sl)
+
+    def pathArray(self,shape,pathwire,count,xlate,align):
+        '''Distribute shapes along a path.'''
+        import Part
+
+        placements = calculatePlacementsOnPath(
+            shape.Placement.Rotation, pathwire, count, xlate, align)
+
+        base = []
+
+        for placement in placements:
+            ns = shape.copy()
+            ns.Placement = placement
+
+            base.append(ns)
+
+        return (Part.makeCompound(base))
+
+
+_PathArray = PathArray
+
+
+def calculatePlacementsOnPath(shapeRotation, pathwire, count, xlate, align):
```

Comment:
I think these functions should not be in this path array module. They are used by it, but seem to be general enough that maybe makes sense to place them in the draftfunctions package.

Reply:
or in the utils? (also if the utils.py module is growing quite big)

Reply:
The way I envision it, I think `utils` should have very small and general functions not dealing with a particular object, while `functions` should have code that deals with more focused actions. For example, `get_type` is very general, because it applies to every single object, but placement on a path is something that we want to calculate only in certain cases, like in the path array, so I think it's not "general" enough.

The module `utils` has various commands that deal with groups, I think these should be moved to a specific `groups` module. That's another way of slimming this down. In general we should try to keep modules at fewer than 1000 lines, and fewer than 500 if possible.

Reply:
well, since at the moment it's used just by path array, i prefer to leave it there: draftfunctions are currently functions that are called by draftguitools. So it's strange to me to move there, into it's own module, a function that is called by another function (i would intuitively not look for it there)...
we can for sure change it later.

Reply:
But there is no reason those functions should only work inside the `draftguitools`. As long as they are provided inside the `Draft.py` module, they should be able to be used from a terminal only session, like `Draft.move`, `Draft.copy`, etc.

Reply:
Sure, `DraftGeomUtils `is another place where this kind of function could go... let's think about that a bit before moving `calculatePlacementsOnPath `to it's own modue :)

Reply:
Actually, `DraftGeomUtils` has to be broken into smaller modules as well, because we don't want to have extremely large libraries. For example, there are many functions related to circles; apparently this is code that was just placed there but is unfinished. So, we can move all these functions to a module that only deals with circles, `draftgeomutils/geo_circles.py` for example.

That's the flexibility of Python. We can place functions in smaller modules, and then bring them all together by importing them.